iis/mymodule.cpp

```diff
@@ -1114,115 +666,90 @@ CMyHttpModule::OnBeginRequest(IHttpContext* httpContext, IHttpEventProvider* pro
-
-        return RQ_NOTIFICATION_FINISH_REQUEST;
-	}
-
-	return RQ_NOTIFICATION_CONTINUE;
-}
-
-apr_status_t ReadResponseCallback(request_rec *r, char *buf, unsigned int length, unsigned int *readcnt, int *is_eos)
-{
-    RequestStoredContext* rsc = RetrieveIISContext(r);
-
-    *readcnt = 0;
-
-    if(rsc == nullptr || rsc->responseBuffer == nullptr)
+    if (config->config->is_enabled == MODSEC_DETECTION_ONLY)
     {
-        *is_eos = 1;
-        return APR_SUCCESS;
-    }
-
-    unsigned int size = length;
-
-    if(size > rsc->responseLength - rsc->responsePosition)
-        size = rsc->responseLength - rsc->responsePosition;
-
-    memcpy(buf, rsc->responseBuffer + rsc->responsePosition, size);
-
-    *readcnt = size;
-    rsc->responsePosition += size;
-
-    if(rsc->responsePosition >= rsc->responseLength)
-        *is_eos = 1;
-
-    return APR_SUCCESS;
-}
-
-apr_status_t WriteResponseCallback(request_rec *r, char *buf, unsigned int length)
-{
-    RequestStoredContext* rsc = RetrieveIISContext(r);
-
-    if(!rsc || !rsc->requestRec || rsc->responseBuffer == nullptr)
-        return APR_SUCCESS;
-
-    IHttpContext *pHttpContext = rsc->httpContext;
-    IHttpResponse *pHttpResponse = pHttpContext->GetResponse();
-    HTTP_RESPONSE *pRawHttpResponse = pHttpResponse->GetRawHttpResponse();
-    HTTP_DATA_CHUNK *pDataChunk = (HTTP_DATA_CHUNK *)apr_palloc(rsc->requestRec->pool, sizeof(HTTP_DATA_CHUNK));
-
-    pRawHttpResponse->EntityChunkCount = 0;
-
-    // since we clean the APR pool at the end of OnSendRequest, we must get IIS-managed memory chunk
-    //
-    void *reqbuf = pHttpContext->AllocateRequestMemory(length);
-
-    memcpy(reqbuf, buf, length);
-
-    pDataChunk->DataChunkType = HttpDataChunkFromMemory;
-    pDataChunk->FromMemory.pBuffer = reqbuf;
-    pDataChunk->FromMemory.BufferLength = length;
+        // This could have been a simple lambda with generalized capture but C++11 doesn't support it yet.
+        // Can be replaced/simplified in C++14.
+        class BackgroundRequestProcessingTask
+        {
+        private:
+            AprRequestContext ctx;
+        public:
+            explicit BackgroundRequestProcessingTask(AprRequestContext&& ctx)
+                : ctx(std::move(ctx))
+            {}
+
+            BackgroundRequestProcessingTask(const BackgroundRequestProcessingTask&) = delete;
+            BackgroundRequestProcessingTask& operator=(const BackgroundRequestProcessingTask&) = delete;
+
+            BackgroundRequestProcessingTask(BackgroundRequestProcessingTask&& rhs)
+                : ctx(std::move(rhs.ctx))
+            {
+            }
 
-    CHAR szLength[21]; //Max length for a 64 bit int is 20
+            BackgroundRequestProcessingTask& operator=(BackgroundRequestProcessingTask&& rhs)
+            {
+                ctx = std::move(rhs.ctx);
+                return *this;
+            }
 
-    ZeroMemory(szLength, sizeof(szLength));
+            int operator()() const {
```

Comment:
does this mean "overload the () on BackgroundRequestProcessingTask?"? How is it different from a constructor?

Reply:
This defines an `operator()` on `BackgroundRequestProcessingTask` so that the instance of this class becomes invocable with the specified set of parameters (in this case - an empty parameters list).

Essentially, C++ lambdas are nothing more than syntastic sugar over declaring some anonymous structs/classes that have `operator()` overloaded for the specified input parameters. Here I'm doing the same, just explicitly, for VS version we are building ModSecurity with lacks support for C++14 generic capture lists with lambdas and, as a consequence, cannot capture movable but non-copyable parameters like `std::unique_ptr`.

Reply:
See more on the subject here: https://en.cppreference.com/w/cpp/language/operators#Function_call_operator